lstm.go

```diff
@@ -2,39 +2,220 @@ package golgi
 
 import (
 	"github.com/chewxy/hm"
+	"github.com/pkg/errors"
 	G "gorgonia.org/gorgonia"
 	"gorgonia.org/tensor"
 )
 
-// LSTM represents a static LSTM model
-type LSTM struct{}
+// LSTM represents an LSTM RNN
+type LSTM struct {
+	name string
+
+	g *G.ExprGraph
+
+	input  lstmGate
+	forget lstmGate
+	output lstmGate
+	cell   lstmGate
+
+	size        int // for construction
+	initialized bool
+	dummyCell   *G.Node
+	dummyHidden *G.Node
+}
+
+// FromLSTMData will initialize a new LSTM model
+func FromLSTMData(g *G.ExprGraph, layer *LSTMData, name string) *LSTM {
+	retVal, err := layer.Make(g, name)
+	if err != nil {
+		panic(err)
+	}
+	return retVal.(*LSTM)
+
+}
+
+// ConsLSTM is a LSTM construction function. It takes a gorgonia.Input that has a *gorgonia.Node.
+func ConsLSTM(in G.Input, opts ...ConsOpt) (retVal Layer, err error) {
+	x := in.Node()
+	if x == nil {
+		return nil, errors.Errorf("LSTM expects a *Node. Got input %v of  %T instead", in, in)
+	}
+
+	// TODO: Ensure shape is being set correctly
+	inshape := x.Shape()
+	if inshape.Dims() > 2 || inshape.Dims() == 0 {
+		return nil, errors.Errorf("Expected shape is either a vector or a matrix")
+	}
+
+	l := &LSTM{}
+	for _, opt := range opts {
+		var (
+			o  Layer
+			ok bool
+		)
+
+		if o, err = opt(l); err != nil {
+			return
+		}
+
+		if l, ok = o.(*LSTM); !ok {
+			err = errors.Errorf("Construction Option returned a non LSTM. Got %T instead", o)
+			return
+		}
+	}
+
+	if err = l.Init(x); err != nil {
+		return
+	}
+
+	retVal = l
+	return
+}
 
 // Model will return the gorgonia.Nodes associated with this LSTM
 func (l *LSTM) Model() G.Nodes {
-	panic("not implemented")
+	return G.Nodes{
+		l.input.wx, l.input.wh, l.input.b,
+		l.forget.wx, l.forget.wh, l.forget.b,
+		l.output.wx, l.output.wh, l.output.b,
+		l.cell.wx, l.cell.wh, l.cell.b,
+	}
 }
 
-// Fwd runs the equation forwards
+// Fwd runs the equation forwards.
+//
+// While a *LSTM can take any gorgonia.Input as an input, it returns a gorgonia.Result,
+// of which the concrete type is a lstimIO.
+//
+// The lstmIO type is not exported. Instead, to query the *Node of the gorgonia.Input or gorgonia.Result,
+// use the Nodes() method.
+//
+// The Result will always be organized as such: [previousHidden, previousCell]
+//
+// e.g.
+// 	out := lstm.Fwd(x)
+// 	outNodes := out.Nodes()
+//	prevHidden := outNodes[0]
+//	prevCell := outNodes[1]
```

Comment:
It also returns the (prev?) input vector.

Reply:
It should not.

Reply:
Do I need to update something? These are all commented out.

Reply:
Nah. It's all good. Waiting on you to confirm you have nothing else to put into this PR